F# .NET Core Template

slang25 opened the pull request:
Implements #3044

## How did you implement it:
I followed a similar set of changes that were introduced to support the C# template, and adapted it for F#.

## Todos:

- [x] Add tests
- [ ] Investigate integration tests

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

lukemerrett commented on lib/plugins/create/templates/aws-fsharp/build.sh:
```diff
@@ -0,0 +1,14 @@
+apt-get -qq -y install zip
+
+#create deployment package
+pushd bin/Release/netcoreapp1.0/publish
```
@slang25 I wonder, rather than lines 7-14, could we instead use:

```
dotnet lambda package --configuration Release --framework netcoreapp1.0
```

As far as I can see the zip file it creates is just the /publish folder anyway

slang25 replied:
Good suggestion, what name does the zip file get?

lukemerrett replied:
By default it's the same name as the folder you're running the script from.  Override with the `--output-package bin/Release/netcoreapp1.0/Package.zip` flag, e.g:

```
dotnet lambda package --configuration Release --framework netcoreapp1.0  --output-package bin/Release/netcoreapp1.0/Package.zip
```

lukemerrett replied:
I believe, bar a readme file, this is good to go, for ref @slang25 : https://github.com/lukemerrett/FSharp-Template-for-Aws-Lambda/tree/serverless/serverless

slang25 replied:
I think we should name the zip `deploy-package.zip` for consistancy with the other templates

pmuens commented on docs/providers/aws/examples/hello-world/fsharp/serverless.yml:
```diff
@@ -0,0 +1,84 @@
+
+# you can add packaging information here
+package:
+ artifact: bin/Release/netcoreapp1.0/publish/deploy-package.zip
```
On my system it's `bin/release/netcoreapp1.0/deploy-package.zip` (lowercase and not in `publish` sub directory).

I built it with the `build.sh` script.

pmuens commented on docs/providers/aws/examples/hello-world/fsharp/serverless.yml:
```diff
@@ -0,0 +1,84 @@
+# Welcome to Serverless!
```
Can you please add a `README.md` file for this example like it's done with the C# example:

https://github.com/serverless/serverless/blob/master/docs/providers/aws/examples/hello-world/csharp/README.md

pmuens replied:
Resolved... Thanks!

pmuens commented on lib/plugins/create/templates/aws-fsharp/serverless.yml:
```diff
@@ -0,0 +1,84 @@
+#      - s3: ${env:BUCKET}
+#      - schedule: rate(10 minutes)
+#      - sns: greeter-topic
+#      - stream: arn:aws:dynamodb:region:XXXXXX:table/foo/stream/1970-01-01T00:00:00.000
```
Resolve... Thanks!

pmuens commented on docs/providers/aws/examples/hello-world/fsharp/serverless.yml:
```diff
@@ -0,0 +1,84 @@
+#      - s3: ${env:BUCKET}
+#      - schedule: rate(10 minutes)
+#      - sns: greeter-topic
+#      - stream: arn:aws:dynamodb:region:XXXXXX:table/foo/stream/1970-01-01T00:00:00.000
```
Resolved... Thanks!

slang25 commented:
Thanks for the feedback. Regarding the `LambdaException`, I'm not sure at the moment. Me and @lukemerrett are going to look at this together in an hour and will investigate.

pmuens commented:
Great! Thanks @slang25 👍

eahefnawy commented:
@slang25  we're releasing v1.5 today. Unfortunately this won't make it in. Thanks for your great work, I'm certain this will be included in the next release once merged 😊

slang25 commented:
@pmuens I have addressed your feedback, thanks for your help.

We investigated OSX and linux support and spotted that the packaging of the zip was failing, this turned out to be a bug in `Amazon.Lambda.Tools` with relative paths and is almost certainly why in the csharp variant this wasn't used. This is now fixed in the version we've updated to.

If you try again, I think it will work for you. If not then let us know 🙂

pmuens commented:
Hey @slang25 thanks for the updates and the note about the `Lambda.Tools` update.

I tested it again today. Unfortunately I still get the following error:

```
Unable to load assembly 'FsharpHandlers, Culture=neutral, PublicKeyToken=null'.
```

I pushed some minor fixes on top of your PR (one updates the Lambda tools for F# as well). Unfortunately the `Lambda.Tools` update seems to not resolve the issue...

Do you maybe have another idea what might be missing here? Thanks in advance and thanks for your work on this PR!

slang25 commented:
@pmuens I have tried building, and deploying with Windows and macOS, both have worked for me. I'm not sure what the problem could be, what OS are you deploying from? And could you send me the zip package?
Thanks

pmuens commented:
Hey @slang25 thanks for the response.

Here are the steps I did to build and deploy the F# service (basically the same I did with the C# service). The whole service is attached below:


1. Checking out this PR
2. Spinning up the `serverless-node` Docker container
3. Creating a new `aws-fsharp` service with the `serverless create` command
4. Exit the `serverless-node` Docker container
5. Spin up the `aws-fsharp` Docker container to build it with the `build.sh` script (Changed permissions to `755` for `build.sh` before that)
6. Exit the `aws-fsharp` Docker container
7. Spin up the `serverless-node` container and deploy the service
8. Run `serverless invoke -f hello` which brings up the following error message:

```
{
    "errorType": "LambdaException",
    "errorMessage": "Unable to load assembly 'FsharpHandlers, Culture=neutral, PublicKeyToken=null'."
}
```

[fsharp.zip](https://github.com/serverless/serverless/files/732024/fsharp.zip)

pmuens commented:
Hey @slang25 did you already had time to debug this / do you need anything else we could do to help you here?

pmuens commented:
Any updates on this one @slang25 ?

jimmyp commented:
Did anyone discover the problem with this? Sounds like a dotnet core version issue?

pmuens commented:
Just a quick update that we've merged https://github.com/serverless/serverless/pull/3065 recently and the `.gitignore` file should be renamed to `gitignore` (w/o a dot in the beginning). Thanks!

slang25 commented:
Sorry all, I've been juggling a few things recently, I'll get back on this tonight 😃

pmuens commented:
Sounds good! No worries / hurries @slang25 👍

jimmyp commented:
Need any help with this @slang25 ?

jimmyp commented:
Awesome!

slang25 commented:
Ok @pmuens, I've deployed and invoked from windows and it looks good. Could you please review.

My only worry is that I haven't verified `build.sh`. If someone has a mac could they follow the instructions here: https://github.com/serverless/serverless/pull/3097#issuecomment-276088579
But with `aws-csharp` replaced with `aws-fsharp` of course.

Failing that, I can grab a mac on Friday and try it out. My mac here has died.

jimmyp commented:
I can do that for you in the next few hours
On Thu, 23 Mar 2017 at 10:25 am, slang25 <notifications@github.com> wrote:

> Ok @pmuens <https://github.com/pmuens>, I've deployed and invoked from
> windows and it looks good. Could you please review.
>
> My only worry is that I haven't verified build.sh. If someone has a mac
> could they follow the instructions here: #3097 (comment)
> <https://github.com/serverless/serverless/pull/3097#issuecomment-276088579>
> But with aws-csharp replaced with aws-fsharp of course.
>
> Failing that, I can grab a mac on Friday and try it out. My mac here has
> died.
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/3045#issuecomment-288570542>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALOZEU0IxhuPy6RcO69h6OSn4XuImDwks5roa3fgaJpZM4LZPhw>
> .
>

jimmyp commented:
Sorry I thought I could easily help with this but I get stuck around `9. ran chmod +x ./build.sh` where the instructions dont make sense to me anymore.

pmuens commented:
Thanks for the update @slang25 👍 

Just added a quick update to the `serverless.yml` so that it's in sync with the other templates. Unfortunately I get the following error when creating a new template and trying to build it (with `./build.sh`) on Linux (in a Docker container):

```
root@e3ab49ff10b7:/app/tmpdirs-serverless/A-Services/fsharp# ./build.sh 
warn : The folder '/app/tmpdirs-serverless/A-Services/fsharp' does not contain a project to restore.
No executable found matching command "dotnet-lambda"
```

Am I doing something wrong? Thanks in advance! We're almost there!

slang25 commented:
@pmuens ah, I had the old version of the docker image referenced.

slang25 commented:
Thanks, I'll do some investigate today, should be able to borrow a Mac

On Fri, 24 Mar 2017, 8:47 am Philipp Muens, <notifications@github.com>
wrote:

> *@pmuens* requested changes on this pull request.
>
> Thanks for updating @slang25 <https://github.com/slang25> 👍
>
> We're almost there. Unfortunately I still get the old error message that
> the assembly cannot be loaded:
>
> {
>     "errorType": "LambdaException",
>     "errorMessage": "Unable to load assembly 'FsharpHandlers, Culture=neutral, PublicKeyToken=null'."
> }
>
> I've used the build.sh file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/serverless/serverless/pull/3045#pullrequestreview-28852684>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABR4BjiuU24aI_i7szJrmSAWLSx6CDKlks5ro4MLgaJpZM4LZPhw>
> .
>

pmuens commented:
Thanks @slang25 👍 

Just drop a line if you need some help!

pmuens commented:
@slang25 thanks again for getting this started!

Just a friendly ping. Any update on this one? Do you need any help here?

I think this goes hand-in-hand with #3391 and should be resolved once #3391 is fixed.

eahefnawy commented:
@slang25 thanks for your work here. I'm closing this PR as it seems that it's now dead without any activity/updates from your side.

I've updated our contribution guidelines to highlight this

Thanks again 😊
